### PR TITLE
add service tests with postgres

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: <Link to Jira ticket>
<Put notes here to help reviewer understand this PR>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 1]
```diff
@@ -397,14 +404,13 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
-    val groupId = samsql"""delete from ${GroupMemberTable as gm}
-        using ${PolicyTable as p}
-        where ${gm.groupId} = ${p.groupId}
-        and ${p.name} = ${id.accessPolicyName}
-        and ${p.resourceId} = (${ResourceTable.loadResourcePK(id.resource)}) returning ${p.groupId}""".updateAndReturnGeneratedKey().apply()
+    val groupId = samsql"${workbenchGroupIdentityToGroupPK(id)}".map(rs => rs.long(1)).single().apply().getOrElse {
+      throw new WorkbenchException(s"Group for policy [$id] not found")
+    }
+
+    samsql"delete from ${GroupMemberTable as gm} where ${gm.groupId} = ${groupId}".update().apply()
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:411]: The old code was nice in that it deleted and returned the id in one round trip but it does not work when there are no members in the policy to start with. I reformulated select the id first, then delete. I thought this was more readable and reliable than handling an exception and doing a second query only on exception.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 2]
```diff
@@ -437,11 +443,11 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
...
-    val policyPK = PolicyPK(
-      samsql"""delete from ${PolicyActionTable as pa}
-              using ${PolicyTable as p}
-              where ${p.name} = ${id.accessPolicyName}
-              and ${p.resourceId} = (${ResourceTable.loadResourcePK(id.resource)}) returning ${p.id}""".updateAndReturnGeneratedKey().apply())
+    val policyPK = samsql"select ${p.id} from ${PolicyTable as p} where ${p.name} = ${id.accessPolicyName} and ${p.resourceId} = (${ResourceTable.loadResourcePK(id.resource)})".map(rs => PolicyPK(rs.long(1))).single().apply().getOrElse {
+      throw new WorkbenchException(s"policy record not found for $id")
+    }
+
+    samsql"delete from ${PolicyActionTable as pa} where ${pa.resourcePolicyId} = $policyPK".update().apply()
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:440]: Same comment as in `overwritePolicyMembersInternal`. Also I think the old code deletes all records from `PolicyActionTable` 😱 (I don't see a constraint on `pa.resourcePolicyId`)

Comment (marctalbott) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala:440]: Looks good, could you add the missing constraint on the delete query in `overwritePolicyRolesInternal`? 
```
samsql"""delete from ${PolicyRoleTable as pr}
using ${PolicyTable as p}
where ${p.name} = ${id.accessPolicyName}
and ${p.resourceId} = (${ResourceTable.loadResourcePK(id.resource)})
and ${pr.resourcePolicyId} = ${p.id}
returning ${p.id}""".updateAndReturnGeneratedKey().apply())```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 3]
```diff
@@ -145,10 +158,16 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
+      Try {
+        // foreign keys in accessInstructions and groupMember tables are set to cascade delete
+        // note: this will not remove this group from any parent groups and will throw a
+        // foreign key constraint violation error if group is still a member of any parent groups
+        samsql"delete from ${GroupTable as g} where ${g.name} = ${groupName}".update().apply()
+      }.recoverWith {
+        case fkViolation: PSQLException if fkViolation.getSQLState == PSQLStateExtensions.FOREIGN_KEY_VIOLATION =>
+          Failure(new WorkbenchExceptionWithErrorReport(
+            ErrorReport(StatusCodes.Conflict, s"group ${groupName.value} cannot be deleted because it is a member of at least 1 other group")))
+      }.get
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:170]: Policies have groups - is it ever possible for a policy's group to be deleted under it?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 4]
```diff
@@ -740,13 +773,23 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
-      accessInstructionsOpt.map(_.instructions)
+
+      accessInstructionsOpt match {
+        case None =>
+          // the group does not exist
+          throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"$groupName not found"))
+        case Some(accessInstructionsRecord) =>
+          // the group exists but the access instructions will be null in the query result if there are no instructions
+          Option(accessInstructionsRecord.instructions)
+      }
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:794]: 👍

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorServiceSpec.scala [hunk 5]
```diff
@@ -127,13 +127,31 @@ class PolicyEvaluatorServiceSpec extends AsyncFlatSpec with Matchers with TestSu
 
   def setup(): IO[Unit] = {
     for{
+      _ <- clearDatabase()
+      _ <- dirDAO.createUser(WorkbenchUser(dummyUserInfo.userId, Some(TestSupport.genGoogleSubjectId()), dummyUserInfo.userEmail))
+    } yield ()
+  }
```
Comment (ansingh7115) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorServiceSpec.scala:133]: is there a reason this is not in a `before {...}` block like in other Spec classes?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala (hunk 3 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:170]: there is a foreign key constraint to prevent that

Comment (ansingh7115) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:170]: oh duh

**Pull request merged**